repository/src/main/java/org/alfresco/repo/action/access/ActionAccessRestrictionAbstractBase.java

```diff
@@ -0,0 +1,156 @@
+ * %%
+ * This file is part of the Alfresco software.
+ * If the software was purchased under a paid Alfresco license, the terms of
+ * the paid license agreement will prevail.  Otherwise, the software is
+ * provided under the following open source license terms:
+ *
+ * Alfresco is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+
+package org.alfresco.repo.action.access;
+
+import org.alfresco.repo.action.ActionModel;
+import org.alfresco.repo.rule.RuleModel;
+import org.alfresco.service.cmr.action.Action;
+import org.alfresco.service.cmr.repository.NodeRef;
+import org.alfresco.service.cmr.repository.NodeService;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class ActionAccessRestrictionAbstractBase implements ActionAccessRestriction {
+
+    private static final List<String> CONTROLLED_ACTION_ACCESS_CONTEXT =
+            List.of(ActionAccessRestriction.RULE_ACTION_CONTEXT, ActionAccessRestriction.FORM_PROCESSOR_ACTION_CONTEXT,
+                    ActionAccessRestriction.V0_ACTION_CONTEXT, ActionAccessRestriction.V1_ACTION_CONTEXT);
+
+    protected NodeService nodeService;
+    private Properties configProperties;
+
+
+    public void setNodeService(NodeService nodeService) {
+        this.nodeService = nodeService;
+    }
+
+    public void setConfigProperties(Properties configProperties) {
+        this.configProperties = configProperties;
+    }
+
+    /**
+     * Base for verifying access restriction,
+     * manages common checks for exposing action in config or action being ran as a consequence of running a rule (safe)
+     *
+     * @param action
+     */
+    public void verifyAccessRestriction(Action action) {
+        if (isActionExposed(action) || isActionCausedByRule(action)) {
+            return;
+        }
+
+        innerVerifyAccessRestriction(action);
+    }
+
+    private boolean isActionExposed(Action action) {
+        return !isActionFromControlledContext(action) || isExposedInConfig(action).orElse(Boolean.FALSE);
+        //TODO consider idea of action from not controlled contexts but restricted in config
+        //e.g. new, custom customer action context
+        //I don't think it has that much merit though
```

Comment:
do you still want this TODO?

Reply:
+1

Reply:
I need to talk to Piotr about it, there is plenty of possible customisation to be done through config, I just don't know if it's needed

Reply:
What we have is enough. It is possible to disable the restriction check in all contexts or in a specific one.

Reply:
Removing TODO then